File: SearchInRotatedSortedArrayII.java
/**
 * LeetCode 81 Search in Rotated Sorted Array II
 */
class Solution {
    public boolean search(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return false;
        }
        int lo = 0;
        int hi = nums.length - 1;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            if (nums[mid] == target) {
                return true;
            }
            if (nums[mid] == nums[lo]) {
                lo++;
                continue;
            }
            if (nums[mid] > nums[lo]) {
                if (nums[lo] <= target && target < nums[mid]) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            } else {
                if (nums[mid] < target && target <= nums[hi]) {
                    lo = mid + 1;
                } else {
                    hi = mid - 1;
                }
            }
        }
        return false;
    }
}

File: SearchInRotatedSortedArrayII.go
func search(nums []int, target int) bool {
	n := len(nums)
	l := 0
	r := n - 1
	for (l <= r) {
		mid := (l + r) / 2
		if (nums[mid] == target) {
			return true
		}
		if (nums[l] == nums[mid]) {
			l++
		} else if (nums[mid] <= nums[r]) {
			if (target > nums[mid] && target <= nums[r]) {
				l = mid + 1
			} else {
				r = mid - 1
			}
		} else {
			if (target >= nums[l] && target < nums[mid]) {
				r = mid - 1
			} else {
				l = mid + 1
			}
		}
	}
	return false
}
